### PR TITLE
Cosine similarity model

### PR DESCRIPTION
**Pull request by madisonmay**

### CODE REVIEW

**Comment by madisonmay** [comment]
Just using this as a diff viewer for now :)

File: finetune/network_modules.py [hunk 1]
```diff
@@ -312,3 +312,38 @@ def seq_lab_internal(hidden):
...
+    :param reuse: Should reuse be set within this scope.
+    :param kwargs: Spare arguments.
+    :return: dict containing:
+        logits: The unnormalised log probabilities of each class.
+        losses: The loss for the classifier.
+    """
+    with tf.variable_scope('cosine_similarity', reuse=reuse):
+        hidden1 = dropout(hidden1, config.clf_p_drop, train, dropout_placeholder)
+        hidden2 = dropout(hidden2, config.clf_p_drop, train, dropout_placeholder)
+        cos_sim_logits = tf.sigmoid(tf.multiply(hidden1, hidden2))
```
Comment (madisonmay) [inline finetune/network_modules.py]: I don't think this is the loss function typically used in Siamese networks.  You might want to take a look at the contrastive loss function from the [tensorflow examples](https://github.com/tiagofrepereira2012/examples.tensorflow/blob/master/examples/tensorflow/script/train_mnist_siamese.py) or from this [hackernoon article](https://hackernoon.com/one-shot-learning-with-siamese-networks-in-pytorch-8ddaab10340e).

**Comment by madisonmay** [comment]
@chaitjo Your life will be easier if you inherit from the [Comparison class](https://github.com/IndicoDataSolutions/finetune/blob/development/finetune/comparison.py#L9) and pretty much change only the `_text_to_ids`, `define_placeholders` and `target_model` methods.

In the `_text_to_ids` method, you stack the token_ids on axis 1.
In the `_define_placeholder` method, you set up placeholders to expect an additional axis.

This causes the finetune featurizer to broadcast operations over that dimension and "featurize" the two document independently in the same manner you do in your branch currently.  

Finally, instead of calling `super()._target_model()` in the target model method, you can use your own custom `target_model` that produces siamese net losses from a feature vector of shape (batch_size, 2, sequence_length, hidden_size)

Does that make sense @chaitjo?

**Comment by madisonmay** [comment]
It's entirely possible I'm missing something, though, let me know if you think there is necessary hidden complexity that I'm missing.

File: finetune/siamese_comparison.py [hunk 2]
```diff
@@ -0,0 +1,544 @@
...
+                    X1,
+                    config=self.config,
+                    encoder=self.encoder,
+                    dropout_placeholder=self.do_dropout,
+                    train=train,
+                    reuse=do_reuse
+                )
+
+                # TODO what does the reuse parameter do here?
+                # Setting `reuse=1` here assuming it means the siamese networks share weights...
```
Comment (madisonmay) [inline finetune/siamese_comparison.py]: This is the reuse parameter that gets passed to `tf.variable_scope()` -- see https://www.tensorflow.org/guide/variables.

File: finetune/network_modules.py (hunk 1 above)
Comment (chaitjo) [reply finetune/network_modules.py]: Technically, you're right. Since I had originally intended to use `finetune` for my own research, I had a hack-y way of implementing the contrastive loss with negative samples using binary cross entropy.   Emailing you about this.

**Comment by chaitjo** [comment]
Created a new branch based on the suggestions: https://github.com/chaitjo/finetune/tree/siamese_comparison_new

I think I understand how featurizer and base model much better now! 😃

**Comment by madisonmay** [comment]
Closing this since you have a new branch up!  Opened a PR for that current iteration at #144 -- again mostly just so I can use github's handy diff viewer, haha.
